include/deal.II/sundials/n_vector.templates.h

```diff
@@ -685,8 +693,104 @@ SUNDIALS::internal::NVectorOperations::weighted_rms_norm_mask(N_Vector x,
   return tmp.l2_norm() / std::sqrt(n);
 }
 
+template <typename VectorType>
+booleantype
+SUNDIALS::internal::NVectorOperations::constraint_mask(N_Vector c,
+                                                       N_Vector x,
+                                                       N_Vector mask)
+{
+  // TODO copy can be avoided by a custom kernel
+  auto *c_dealii      = unwrap_nvector_const<VectorType>(c);
+  auto *x_dealii      = unwrap_nvector_const<VectorType>(x);
+  auto *mask_dealii   = unwrap_nvector<VectorType>(mask);
+  bool  everything_ok = true;
+
+  AssertDimension(c_dealii->size(), x_dealii->size());
+  AssertDimension(c_dealii->size(), mask_dealii->size());
+  for (const auto &i : x_dealii->locally_owned_elements())
+    {
+      if ((*c_dealii)[i] == 1)
+        {
+          if ((*x_dealii)[i] >= 0)
+            {
+              (*mask_dealii)[i] = 0.0;
+            }
+          else
+            {
+              (*mask_dealii)[i] = 1.0;
+              everything_ok     = false;
+            }
+        }
+      else if ((*c_dealii)[i] == 2)
+        {
+          if ((*x_dealii)[i] > 0)
+            {
+              (*mask_dealii)[i] = 0.0;
+            }
+          else
+            {
+              (*mask_dealii)[i] = 1.0;
+              everything_ok     = false;
+            }
+        }
+      else if ((*c_dealii)[i] == -1)
+        {
+          if ((*x_dealii)[i] <= 0)
+            {
+              (*mask_dealii)[i] = 0.0;
+            }
+          else
+            {
+              (*mask_dealii)[i] = 1.0;
+              everything_ok     = false;
+            }
+        }
+      else if ((*c_dealii)[i] == -2)
+        {
+          if ((*x_dealii)[i] < 0)
+            {
+              (*mask_dealii)[i] = 0.0;
+            }
+          else
+            {
+              (*mask_dealii)[i] = 1.0;
+              everything_ok     = false;
+            }
+        }
+      else
+        {
+          Assert((*c_dealii)[i] == 0,
+                 ExcMessage("KINSOL constraints value is not admissible"));
+          (*mask_dealii)[i] = 0.0;
+        }
+    }
+  if (everything_ok)
+    return SUNTRUE;
+  else
+    return SUNFALSE;
+}
 
 
+template <typename VectorType>
+realtype
+SUNDIALS::internal::NVectorOperations::min_quotient(N_Vector n, N_Vector d)
```

Comment:
same here, it would be nice to have a high-level description of what this function does